# Comparative Exploratory Data Analysis

This notebook contains a comparative exploratory analysis of the two datasets used in this project.

The aim of this notebook is to highlight the main characteristics of the original dataset by exploring its variables and statistical components.

The second aim of this notebook is to generate a synthetic dataset from the original dataset, so that its characteristics can be analysed by comparing it with the original dataset.

By analysing the two datasets, we can initially determine whether the synthetic dataset is statistically faithful to the original; this provides us with a great deal of information for the next phase, namely training machine learning models on the synthetic dataset and then testing them on the real dataset.

## Basic setup:

In [1]:
import sys
from pathlib import Path

# add the root folder to the python path 
# (this fixes problems with importing modules inside the notebook)
PROJECT_ROOT = Path().resolve().parent
sys.path.append(str(PROJECT_ROOT))

We prepare the data we are going to use, both the real and the synthetic data.

In [2]:
import sys
import os
from src.data_loader import load_data
from src.eda_utils import plot_hist, plot_bar, plot_box, calculate_statistics
from src.eda_utils import frequency_table, statistical_tests, numeric_biv_analysis
from src.eda_utils import qualitative_association, plot_corr_matrix
from src.synthesizer import generate_syn_df, diagnostic_report, quality_report, get_column_plot

# load original dataset 
df = load_data()
print(df.head())

# generate synthetic data
syn_df = generate_syn_df(df, n_epochs=300)
print(syn_df.head())

   Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0  


c:\Users\Riccardo\miniconda3\envs\project_env\Lib\site-packages\sdv\single_table\base.py:134: UserWarning: We strongly recommend saving the metadata using 'save_to_json' for replicability in future SDV versions.
  warnings.warn(
Gen. (-00.98) | Discrim. (+00.08): 100%|██████████| 300/300 [00:19<00:00, 15.57it/s]


   Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   65   M           ASY        112          238          1     Normal    126   
1   71   M           NAP        162          158          0     Normal    129   
2   36   M           ASY        162          272          0     Normal     85   
3   74   M           ASY        187          311          0     Normal     63   
4   55   F           ASY        121          182          1     Normal     90   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              Y      4.7       Up             0  
1              Y      0.4       Up             0  
2              N      0.1     Flat             1  
3              Y      0.3     Flat             1  
4              Y      0.3     Flat             0  


**Attribute information reported from the dataset Kaggle webpage** :

1. Age: age of the patient (years)
2. Sex: sex of the patient [M: Male, F: Female]
3. ChestPainType: chest pain type [TA: Typical Angina, ATA: Atypical Angina, NAP: Non-Anginal Pain, ASY: Asymptomatic]
4. RestingBP: resting blood pressure [mm Hg]
5. Cholesterol: serum cholesterol [mm/dl]
6. FastingBS: fasting blood sugar [1: if FastingBS > 120 mg/dl, 0: otherwise]
7. RestingECG: resting electrocardiogram results [Normal: Normal, ST: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV), LVH: showing probable or definite left ventricular hypertrophy by Estes' criteria]
8. MaxHR: maximum heart rate achieved [Numeric value between 60 and 202]
9. ExerciseAngina: exercise-induced angina [Y: Yes, N: No]
10. Oldpeak: oldpeak = ST [Numeric value measured in depression]
11. ST_Slope: the slope of the peak exercise ST segment [Up: upsloping, Flat: flat, Down: downsloping]
12. HeartDisease: output class [1: heart disease, 0: Normal]

In [5]:
print(df.info())

print(syn_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             918 non-null    int64  
 1   Sex             918 non-null    object 
 2   ChestPainType   918 non-null    object 
 3   RestingBP       918 non-null    int64  
 4   Cholesterol     918 non-null    int64  
 5   FastingBS       918 non-null    int64  
 6   RestingECG      918 non-null    object 
 7   MaxHR           918 non-null    int64  
 8   ExerciseAngina  918 non-null    object 
 9   Oldpeak         918 non-null    float64
 10  ST_Slope        918 non-null    object 
 11  HeartDisease    918 non-null    int64  
dtypes: float64(1), int64(6), object(5)
memory usage: 86.2+ KB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  

As we can see, the first dataset consists of a total of 918 rows, with each row representing a patient who may or may not have heart disease. The target variable “HeartDisease” is a binary variable, where 0 indicates the absence of heart disease and 1 indicates the presence of heart disease.

We also note that there do not appear to be any missing values in the dataset; however, we must investigate whether there are any duplicate values, as these could pose problems for the future training of a predictive model.

The synthetic dataset, on the other hand, has 10,000 rows, and from this initial analysis, it too appears to have no missing values.

In [ ]:
# check duplicated rows
print(df.duplicated().sum())

# check du

0
